vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_gpio.c

```diff
@@ -0,0 +1,622 @@
+
+/* Logging Includes */
+#include "iot_logging_task.h"
+
+/* Note, this depends on logging task being created and running */
+#if ( IOT_GPIO_LOGGING_ENABLED == 1 )
+    #define IOT_GPIO_MODULE_NAME    "[CommonIO][GPIO]"
+    #define IOT_GPIO_LOGF( format, ... )    vLoggingPrintf( IOT_GPIO_MODULE_NAME format, __VA_ARGS__ )
+    #define IOT_GPIO_LOG( msg )             vLoggingPrintf( IOT_GPIO_MODULE_NAME msg )
+#else
+    #define IOT_GPIO_LOGF( format, ... )
+    #define IOT_GPIO_LOG( msg )
+#endif
+
+
+#define IOT_GPIO_CLOSED    ( ( uint8_t ) 0 )
+#define IOT_GPIO_OPENED    ( ( uint8_t ) 1 )
+
+#define DEFAULT_GPIO_DESCRIPTOR                    \
+    {                                              \
+        .lGpioNumber = -1,                         \
+        .xConfig =                                 \
+        {                                          \
+            .xDirection     = eGpioDirectionInput, \
+            .xOutMode       = eGpioPushPull,       \
+            .lDriveStrength = NRF_GPIO_PIN_S0S1,   \
+            .xPull          = eGpioPullNone,       \
+            .xInterruptMode = eGpioInterruptNone   \
+        },                                         \
+        .xUserCallback = NULL,                     \
+        .pvUserContext = NULL,                     \
+        .ucState = IOT_GPIO_CLOSED                 \
+    }
+
+typedef struct
+{
+    IotGpioDirection_t xDirection;
+    IotGpioOutputMode_t xOutMode;
+    int32_t lDriveStrength;
+    IotGpioPull_t xPull;
+    IotGpioInterrupt_t xInterruptMode;
+} IotGpioConfig_t;
+
+/**
+ * @brief   GPIO descriptor type defined in the source file.
+ */
+typedef struct IotGpioDescriptor
+{
+    int32_t lGpioNumber;
+    IotGpioConfig_t xConfig;
+    IotGpioCallback_t xUserCallback;
+    void * pvUserContext;
+    uint8_t ucState;
+} IotGpioDescriptor_t;
+
+/*
+ * A 1-to-1 mapping set by iot_gpio_config.h. CommonIO Gpio pin index space uses the index i
+ * and the board specific identifier is stored at GpioMap[i]. The mapping is user configured.
+ */
+static const int32_t plGpioMap[] = IOT_COMMON_IO_GPIO_PIN_MAP;
+
+static IotGpioDescriptor_t pxGpioDesc[ IOT_COMMON_IO_GPIO_NUMBER_OF_PINS ];
+
+
+/*---------------------------------------------------------------------------------*
+*                                Private Helpers                                  *
+*---------------------------------------------------------------------------------*/
+
+/*
+ * @brief   Used to validate whether a CommonIO pin index is a valid argument for iot_gpio_open.
+ *          This lGpioNumber will be used to index xGpioDesc and plGpioMap.
+ *
+ * @param[in] lGpioNumber CommonIO pin index to validate
+ *
+ * @return    true if pin is okay to use with iot_gpio_open, else false
+ */
+bool prvIsValidPinIndex( int32_t lGpioNumber )
+{
+    int32_t ulNumberOfMappings = sizeof( plGpioMap ) == 0 ? 0 : sizeof( plGpioMap ) / sizeof( plGpioMap[ 0 ] );
+
+    return ( ulNumberOfMappings > 0 ) &&
```

Comment:
This was intentional, to help prevent accidentally missed entries. 
Statically sized non-VLA struct arrays can accept subset initializations like the following:
```c
plGpioMap[3] = { 
                   BOARD_PIN(34),
                   BOARD_PIN(6) 
               }
``` 
leaving the last gpio pin uninitialized but _supposedly_ mapped.

By using a static VLA for `plGpioMap`, we can check it's length against `IOT_COMMON_IO_GPIO_NUMBER_OF_PINS` to prevent users from accidentally missing a map entry. To supplement that, there's logs that indicate such incomplete maps.